File: sdk/python/src/openwop_client/sse.py
"""
Generator-based SSE consumer for `GET /v1/runs/{runId}/events`. Pure
stdlib — `urllib.request` for HTTP, manual line parsing for SSE.

Synchronous generator: callers iterate with `for event in stream_events(...)`.
Connection is auto-closed when the server closes the stream OR when the
caller breaks out of the loop. Bounded by an absolute timeout so CI
never hangs.

For async usage, callers can wrap this with `asyncio.to_thread(...)`.
An optional `httpx`-based async client may land in a future v1.x release.
"""

from __future__ import annotations

import json
from typing import Any, Iterator, Sequence
from urllib.parse import urlencode
from urllib.request import Request, urlopen

from .types import RunEventDoc, StreamMode


def stream_events(
    base_url: str,
    api_key: str,
    run_id: str,
    *,
    stream_mode: StreamMode | Sequence[StreamMode] | None = None,
    last_event_id: str | None = None,
    timeout_seconds: float = 30.0,
    buffer_ms: int | None = None,
) -> Iterator[RunEventDoc]:
    """Subscribe to a run's SSE event stream and yield decoded events.

    Args:
        base_url:         openwop server base URL (e.g., `https://api.example.com`).
        api_key:          Bearer-style API key.
        run_id:           Run to subscribe to.
        stream_mode:      Single mode (e.g., 'updates') OR an iterable of modes
                          for S4 mixed-mode (e.g., ('updates', 'messages')).
                          Iterables serialize to the canonical comma-separated
                          query (`?streamMode=updates,messages`).
        last_event_id:    Optional `Last-Event-ID` request header for resumption.
        timeout_seconds:  Hard upper bound on the read. Server SHOULD close on
                          terminal events; this catches misbehavior.
        buffer_ms:        S3 batching hint (0..5000). When set, the server
                          batches events for up to N ms; the SDK transparently
                          flattens batched arrays back into per-event yields,
                          so consumers see the same shape as unbuffered streams.

    Yields:
        RunEventDoc for each parseable event. Non-JSON `data:` payloads
        (keep-alive, vendor extensions) are silently skipped. Batched
        events (S3 `event: batch`) are flattened transparently.

    Raises:
        urllib.error.HTTPError: on non-2xx status.
        urllib.error.URLError:  on connection failure.
    """
    base_url = base_url.rstrip("/")
    params: dict[str, str] = {}
    if stream_mode:
        if isinstance(stream_mode, str):
            params["streamMode"] = stream_mode
        else:
            params["streamMode"] = ",".join(stream_mode)
    if buffer_ms is not None:
        params["bufferMs"] = str(buffer_ms)
    qs = "?" + urlencode(params) if params else ""
    url = f"{base_url}/v1/runs/{run_id}/events{qs}"

    headers = {
        "Accept": "text/event-stream",
        "Authorization": f"Bearer {api_key}",
        "Cache-Control": "no-cache",
    }
    if last_event_id:
        headers["Last-Event-ID"] = last_event_id

    req = Request(url, headers=headers, method="GET")
    with urlopen(req, timeout=timeout_seconds) as resp:
        # urlopen raises on non-2xx, so resp.status is always 2xx here.
        pending_event = "message"
        pending_data: list[str] = []
        pending_id: str | None = None

        for raw_line in resp:
            line = raw_line.decode("utf-8", errors="replace").rstrip("\r\n")

            if line == "":
                # Event boundary — flush. Returns a list because S3 batched
                # `event: batch` events fan out into multiple RunEventDocs.
                events = _flush_event(pending_event, pending_data, pending_id)
                pending_event = "message"
                pending_data = []
                pending_id = None
                for ev in events:
                    yield ev
                continue

            if line.startswith(":"):
                # SSE comment / keep-alive — skip.
                continue

            colon = line.find(":")
            if colon == -1:
                field = line
                value = ""
            else:
                field = line[:colon]
                value = line[colon + 1 :]
                if value.startswith(" "):
                    value = value[1:]

            if field == "event":
                pending_event = value
            elif field == "data":
                pending_data.append(value)
            elif field == "id":
                pending_id = value
            # Unknown fields ignored per RFC 8895.

        # Flush any final unterminated event.
        for ev in _flush_event(pending_event, pending_data, pending_id):
            yield ev


def _flush_event(
    event: str, data_lines: list[str], event_id: str | None
) -> list[RunEventDoc]:
    """Decode a buffered SSE event into RunEventDocs.

    Returns:
        - empty list when buffer is empty / non-JSON / malformed (skip).
        - 1-element list for normal `event: <type>` payloads.
        - N-element list when the server batched per S3 — `event: batch`
          with `data:` as a JSON array of RunEventDoc.
    """
    _ = event_id
    if not data_lines:
        return []
    raw = "\n".join(data_lines)
    try:
        parsed: Any = json.loads(raw)
    except json.JSONDecodeError:
        return []

    # S3 batched envelope — `event: batch` carries an array of events.
    if event == "batch" and isinstance(parsed, list):
        out: list[RunEventDoc] = []
        for item in parsed:
            decoded = _decode_event_doc(item)
            if decoded is not None:
                out.append(decoded)
        return out

    if not isinstance(parsed, dict):
        return []
    decoded = _decode_event_doc(parsed)
    return [decoded] if decoded is not None else []


def _decode_event_doc(parsed: Any) -> RunEventDoc | None:
    """Defensive RunEventDoc construction — returns None on missing/
    misshapen required fields. Forward-compat readers tolerate extras.
    """
    if not isinstance(parsed, dict):
        return None
    try:
        return RunEventDoc(
            eventId=str(parsed["eventId"]),
            runId=str(parsed["runId"]),
            type=str(parsed["type"]),
            payload=parsed.get("payload"),
            timestamp=str(parsed["timestamp"]),
            sequence=int(parsed["sequence"]),
            nodeId=parsed.get("nodeId"),
            schemaVersion=parsed.get("schemaVersion"),
            engineVersion=parsed.get("engineVersion"),
            causationId=parsed.get("causationId"),
        )
    except (KeyError, ValueError, TypeError):
        return None

File: sdk/typescript/src/sse.ts
/**
 * SSE consumer for `GET /v1/runs/{runId}/events`. Async-iterable shape so
 * consumers can write `for await (const event of client.runs.events(...))`.
 *
 * Implementation parses event:/data:/id: lines per RFC 8895. Native fetch +
 * ReadableStream — zero third-party deps.
 *
 * Designed to be cancellable: pass an AbortSignal via options, or break out
 * of the for-await loop and the underlying connection is torn down on the
 * next tick.
 */

import type { RunEventDoc, StreamMode } from './types.js';

export interface EventsStreamOptions {
  /**
   * Single mode (e.g., 'updates') OR array of modes (S4 mixed-mode,
   * e.g., ['updates', 'messages']). Arrays serialize to a comma-
   * separated `?streamMode=updates,messages` query.
   */
  readonly streamMode?: StreamMode | readonly StreamMode[];
  readonly lastEventId?: string;
  readonly signal?: AbortSignal;
  /**
   * S3 batching hint. When set, server batches events for up to N ms;
   * the SDK transparently flattens batched arrays back into individual
   * RunEventDoc yields, so consumers see the same per-event surface as
   * unbuffered streams. Range 0..5000.
   */
  readonly bufferMs?: number;
}

export interface EventsStreamContext {
  readonly baseUrl: string;
  readonly apiKey: string;
}

export async function* streamEvents(
  ctx: EventsStreamContext,
  runId: string,
  opts: EventsStreamOptions = {},
): AsyncGenerator<RunEventDoc, void, void> {
  const params = new URLSearchParams();
  if (opts.streamMode) {
    const modeParam: string =
      typeof opts.streamMode === 'string' ? opts.streamMode : opts.streamMode.join(',');
    params.set('streamMode', modeParam);
  }
  if (opts.bufferMs !== undefined) {
    params.set('bufferMs', String(opts.bufferMs));
  }
  const qs = params.toString();
  const url = `${ctx.baseUrl}/v1/runs/${encodeURIComponent(runId)}/events${qs ? `?${qs}` : ''}`;

  const headers: Record<string, string> = {
    Accept: 'text/event-stream',
    Authorization: `Bearer ${ctx.apiKey}`,
    'Cache-Control': 'no-cache',
  };
  if (opts.lastEventId) {
    headers['Last-Event-ID'] = opts.lastEventId;
  }

  const internalAbort = new AbortController();
  const externalSignal = opts.signal;
  if (externalSignal) {
    if (externalSignal.aborted) internalAbort.abort();
    else externalSignal.addEventListener('abort', () => internalAbort.abort(), { once: true });
  }

  const res = await fetch(url, { method: 'GET', headers, signal: internalAbort.signal });
  if (!res.ok || res.body === null) {
    throw new Error(`SSE subscribe failed: HTTP ${res.status}`);
  }

  const reader = res.body.getReader();
  const decoder = new TextDecoder('utf-8');
  let buffer = '';
  let pendingEvent = 'message';
  let pendingData: string[] = [];
  let pendingId: string | null = null;

  /**
   * Flush the buffered event. Returns an array of RunEventDoc:
   *   - 0 elements when the buffer is empty or non-JSON (skip).
   *   - 1 element for a normal `event: <type>` event.
   *   - N elements when the server batched per S3 — `event: batch` with
   *     `data:` as a JSON array of RunEventDoc.
   */
  const flushAndYield = (): RunEventDoc[] => {
    if (pendingData.length === 0) {
      pendingEvent = 'message';
      pendingId = null;
      return [];
    }
    const dataStr = pendingData.join('\n');
    const eventType = pendingEvent;
    pendingEvent = 'message';
    pendingData = [];
    pendingId = null;
    try {
      const parsed = JSON.parse(dataStr) as unknown;
      // S3 batched envelope — `event: batch` carries an array of events.
      if (eventType === 'batch' && Array.isArray(parsed)) {
        return parsed as RunEventDoc[];
      }
      // Normal single-event payload.
      return [parsed as RunEventDoc];
    } catch {
      // Skip non-JSON events (keep-alive payloads, vendor extensions).
      return [];
    }
  };

  try {
    while (true) {
      const { done, value } = await reader.read();
      if (done) break;
      buffer += decoder.decode(value, { stream: true });
      let nlIdx: number;
      while ((nlIdx = buffer.indexOf('\n')) !== -1) {
        const rawLine = buffer.slice(0, nlIdx).replace(/\r$/, '');
        buffer = buffer.slice(nlIdx + 1);

        if (rawLine === '') {
          for (const event of flushAndYield()) yield event;
          continue;
        }
        if (rawLine.startsWith(':')) continue; // keep-alive comment

        const colon = rawLine.indexOf(':');
        const field = colon === -1 ? rawLine : rawLine.slice(0, colon);
        const valueRaw = colon === -1 ? '' : rawLine.slice(colon + 1);
        const fieldValue = valueRaw.startsWith(' ') ? valueRaw.slice(1) : valueRaw;

        switch (field) {
          case 'event':
            pendingEvent = fieldValue;
            break;
          case 'data':
            pendingData.push(fieldValue);
            break;
          case 'id':
            pendingId = fieldValue;
            break;
          default:
            break;
        }
      }
    }
    // Flush any final unterminated event.
    for (const final of flushAndYield()) yield final;
  } finally {
    try {
      reader.releaseLock();
    } catch {
      // best-effort
    }
    if (!internalAbort.signal.aborted) internalAbort.abort();
  }

  // Reference pendingEvent/pendingId so the linter doesn't flag them as
  // unused; they're consumed via flushAndYield's closure but TS can't see
  // that across a generator boundary.
  void pendingEvent;
  void pendingId;
}
